sc-platform/iota-transactional-test-runner documentation (#5335)
iotaledger/iota: crates/iota-transactional-test-runner/syntax.md

> # Syntactic rules for mock network tasks in `iota-transactional-test-runner`

Reviewer: ```suggestion
# Syntactic Rules for Mock Network Tasks in `iota-transactional-test-runner`
```

Transactional tests simulate network operations through the framework exposed in [iota-transactional-test-runner](https://github.com/iotaledger/iota/tree/develop/crates/iota-transactional-test-runner). The framework is actually built on top of the more generic [move-transactional-test-runner](https://github.com/iotaledger/iota/tree/develop/external-crates/move/crates/move-transactional-test-runner).

> This currently used in the following tests:

Reviewer: ```suggestion
This is currently used in the following tests:
```

```
$ cargo tree -i iota-transactional-test-runner
iota-transactional-test-runner v0.1.0 (crates/iota-transactional-test-runner)
[dev-dependencies]
├── iota-adapter-transactional-tests v0.1.0 (crates/iota-adapter-transactional-tests)
├── iota-graphql-e2e-tests v0.1.0 (crates/iota-graphql-e2e-tests)
└── iota-verifier-transactional-tests v0.1.0 (crates/iota-verifier-transactional-tests)
```

> ## Common rules

Reviewer: ```suggestion
## Common Rules
```

> The framework introduces an ad-hoc syntax for defining network related operations/tasks as an extension to `move/mvir` files.

Reviewer: ```suggestion
The framework introduces an ad hoc syntax for defining network-related operations/tasks as an extension to `move/mvir` files.
```

The syntax uses comments with the `//#` prefix to begin blocks of continuous non-empty lines that are eventually used to parse the underlying tasks and any additional `data`. Empty lines define the boundaries of each block. So the basic syntax for all tasks is the following:

```
<empty-line>
//# <task> [OPTIONS]
[<task-data>]
...
<empty-line>
```

For example:

```
                                                                        [empty-line]
//# run-graphql --show-usage --show-headers --show-service-version      [task]
{                                                                       [data]
  checkpoint {                                                          [data]
    sequenceNumber                                                      [data]
  }                                                                     [data]
}                                                                       [data]
                                                                        [empty-line]
```

The syntax rules for the `data` are specific to each task and will be discussed
in the respective sections.

> ### ObjectID rules

Reviewer: ```suggestion
### ObjectID Rules
```

Object identifiers (ObjectID) follow specific conventions that allow referencing objects across different test commands. This section describes how object IDs work, including how they are used in subcommands and programmable transactions (PTBs).

#### Understanding Object Identifiers (object(x,y))

Object identifiers in test files typically take the form:

```
object(x,y)
```

- x: Represents the task number in which the object was created.
- y: Represents the index of the object within that task.

For instance, object(1,0) means:

The object was created in task 1.
It was the first object (0-based index) created within that task.
In `.move` test files, object references are often written as:

```move
//# view-object 1,0
```

Here, 1,0 refers to the object created in task 1, index 0.

> Howewer, the index order can change due to test execution differences, such as:

Reviewer: ```suggestion
However, the index order can change due to test execution differences, such as:
```

- Non-deterministic transaction execution: Some transactions might reorder operations internally, leading to different assignment orders.
- Dynamic object discovery: Unwrapped objects (e.g., from storage) may be assigned new fake IDs later in the test, shifting the enumeration order.

> ##### How IDs Are Assigned:

Reviewer: ```suggestion
##### How IDs Are Assigned
```

task 2 'view-object'. lines 7-7:
Owner: Account Address ( A )
Version: 2
Contents: iota::coin::Coin<iota::iota::IOTA> {id: iota::object::UID {id: iota::object::ID {bytes: fake(1,0)}}, balance: iota::balance::Balance<iota::iota::IOTA> {value: 1000u64}}

task 3 'create-checkpoint'. lines 9-9:
Checkpoint created: 1

task 4 'programmable'. lines 11-12:
mutated: object(0,0)
deleted: object(1,0)
gas summary: computation_cost: 1000000, storage_cost: 988000,  storage_rebate: 1976000, non_refundable_storage_fee: 0
```

Explanation:

In task 1:

- Object object(0,0) is the gas coin created for account address A in Task 0 and this is split.
- Object object(1,0) is created out of the split.

In task 4:

- object(0,0) is mutated.
- object(1,0) is deleted after being merged into object(0,0).

#### Summary

1. object(x,y): References an object created in task x, at index y.
2. view-object x,y: Displays the current state of the object.
3. object(x,y)@version: Specifies a particular version of the object.
4. Objects in PTBs: Used for transfers (TransferObjects), merging (MergeCoins), and execution of transactions.

> ## Supported tasks

Reviewer: ```suggestion
## Supported Tasks
```

B1:
       4: Branch(9)
B2:
       5: MoveLoc[2](ctx#0#0: &mut TxContext)
       6: Pop
       7: LdU64(1)
       8: Abort
B3:
       9: MoveLoc[2](ctx#0#0: &mut TxContext)
       10: Call object::new(&mut TxContext): UID
       11: MoveLoc[1](amount#0#0: u64)
       12: Pack[0](TestCoin)
       13: MoveLoc[0](to#0#0: address)
       14: Call 1transfer::public_transfer<TestCoin>(TestCoin, address)
       15: Ret
}
}
```

#### Options

```
--syntax <SYNTAX>: move syntax type (`source` or `ir`).
```

- The `Move IR` is a low-level intermediate representation that closely mirrors Move bytecode. The Move bytecode defines programs published to the blockchain. What mainly differentiates the Move IR from the bytecode is that names are used as opposed to indexes into pools/tables.
- The `Move source` language is a high-level language that compiles to Move bytecode. It is designed to be a familiar and ergonomic language for developers that provides minimal abstractions over the Move bytecode.

Example of `.mvir` code:

```mvir
module 0x0.m {
    import 0x2.clock;

    public entry yes_clock_ref(l0: &clock.Clock) {
        label l0:
        abort 0;
    }
}
```

### `publish`

> The publish command allows users to publish Move packages to the IOTA network. This command compiles the specified Move package and deploys it to the network, optionally marking it as upgradable.

Reviewer: ```suggestion
The `publish` command allows users to publish Move packages to the IOTA network. This command compiles the specified Move package and deploys it to the network, optionally marking it as upgradable.
```

        let id = object::new(ctx);
        let test_coin = TestCoin { id, amount };
        transfer::public_transfer(test_coin, to);
    }
}

//# run test::transfer::transfer --sender acc1 --gas-price 500 --args @acc2 50

//# view-object 2,0
```

`test::transfer` should have been published already before you can `run` the command execution.

- Runs transfer function.
- `acc1` is the sender.
- `@acc2` is an identifier of recipient address.
- `50` is an amount of tokens to mint.
- The gas price is set to 500.

`.exp` output:

```
processed 4 tasks

init:
acc1: object(0,0), acc2: object(0,1)

task 1 'publish'. lines 3-18:
created: object(1,0)
mutated: object(0,2)
gas summary: computation_cost: 1000000, storage_cost: 5449200,  storage_rebate: 0, non_refundable_storage_fee: 0

task 2 'run'. lines 20-20:
created: object(2,0)
mutated: object(0,0)
gas summary: computation_cost: 500000, storage_cost: 2363600,  storage_rebate: 0, non_refundable_storage_fee: 0

task 3 'view-object'. lines 22-22:
Owner: Account Address ( acc2 )
Version: 2
Contents: test::transfer::TestCoin {id: iota::object::UID {id: iota::object::ID {bytes: fake(2,0)}}, amount: 50u64}
```

### `view-object`

> The `ViewObject` command retrieves and displays the details of a specific object stored on-chain. Objects can be Move resources, packages, or system objects.

Reviewer: ```suggestion
The `view-object` command retrieves and displays the details of a specific object stored on-chain. Objects can be Move resources, packages, or system objects.
```

#### Syntax

```
//# view-object <ID>
```

#### Options

```
<ID>: the ID of the object to be view.
```

> Example:

Reviewer: ```suggestion
#### Example
```

```move
//# init --accounts acc1 acc2 --protocol-version 1 --simulator

//# view-object 0,0
```

`.exp` output:

```
processed 2 tasks

init:
acc1: object(0,0), acc2: object(0,1)

task 1 'view-object'. lines 3-3:
Owner: Account Address ( acc1 )
Version: 1
Contents: iota::coin::Coin<iota::iota::IOTA> {id: iota::object::UID {id: iota::object::ID {bytes: fake(0,0)}}, balance: iota::balance::Balance<iota::iota::IOTA> {value: 300000000000000u64}}
```

### `transfer-object`

> The `TransferObject` subcommand is used to transfer ownership of an object from one account to another.

Reviewer: ```suggestion
The `transfer-object` subcommand is used to transfer ownership of an object from one account to another.
```

Reviewer: This happens in some more down the doc, unless they should be in PascalCase in which case please ignore the previous comments as well

```

#### Options

```
--sender <SENDER> (optional): specifies the sender of the transaction. If omitted, the default account is used.
--gas-budget <GAS_BUDGET> (optional): defines the gas limit for executing the transaction. If omitted, a default gas budget is used.
--gas-price <GAS_PRICE> (optional): specifies the gas price for this transaction. If not set, the default gas price is used.
--dev-inspect (optional): runs the transaction in inspection mode without committing state changes.
--inputs <INPUTS>: a list of input arguments for the transaction. These inputs are passed as parameters to the commands executed in the programmable transaction.
```

Example:

```move
//# init --addresses test=0x0 --accounts acc1 acc2 --protocol-version 1

//# publish --sender acc1
module test::test_coin {
    public struct TestCoin has key, store {
        id: UID,
        amount: u64
    }

    public fun test_coin_mint(amount: u64, ctx: &mut TxContext) : TestCoin {
        let id = object::new(ctx);
        TestCoin { id, amount }
    }
}

//# programmable --sender acc1 --inputs 1000 @acc2
//> test::test_coin::test_coin_mint(Input(0));
//> TransferObjects([Result(0)], Input(1))
```

> Here we're minting the `TestCoin` obj in fly and passing it to the `TransferObjects` ptb command via Result(0).

Reviewer: ```suggestion
Here, we're minting the `TestCoin` obj in fly and passing it to the `TransferObjects` PTB command via Result(0).
```

`.exp` output:

```
processed 3 tasks

init:
acc1: object(0,0), acc2: object(0,1)

task 1 'publish'. lines 3-14:
created: object(1,0)
mutated: object(0,0)
gas summary: computation_cost: 1000000, storage_cost: 5069200,  storage_rebate: 0, non_refundable_storage_fee: 0

task 2 'programmable'. lines 16-18:
created: object(2,0)
mutated: object(0,0)
gas summary: computation_cost: 1000000, storage_cost: 2371200,  storage_rebate: 988000, non_refundable_storage_fee: 0
```

The `programmable` subcommand is constructed using the same input, result and command components of a Programmable Transaction Block (PTB).

Inputs are the values you provide to the PTB, either as objects or pure values, while Results are the values produced by the commands within the PTB:

- `Input(u16)`: is an input argument, where the `u16` is the index of the input in the input vector. For example, given an input vector of `[Object1, Object2, Object3, Object4]`, `Object1` is accessed with `Input(0)` and `Object3` is accessed with `Input(2)`.

> - `Gas`: is a special input argument representing the object for the `IOTA` coin used to pay for gas. It is kept separate from the other inputs because the gas coin is always present in each transaction and has special restrictions (you can only use it by-value with the `TransferObjects` command) not present for other inputs. Additionally, the gas coin being separate makes its usage explicit, which is helpful for sponsored transactions where the sponsor might not want the sender to use the gas coin for anything other than gas.

Reviewer: ```suggestion
- `Gas`: is a special input argument representing the object for the `IOTA` coin used to pay for gas. It is kept separate from the other inputs because the gas coin is always present in each transaction and has special restrictions (you can only use it by-value with the `TransferObjects` command) that are not present for other inputs. Additionally, the gas coin being separate makes its usage explicit, which is helpful for sponsored transactions where the sponsor might not want the sender to use the gas coin for anything other than gas.
```

The `UpgradePackage` subcommand is used to upgrade an existing Move package on-chain. This allows for adding new features, fixing bugs, or optimizing performance while maintaining compatibility with previous versions.

#### Syntax

```
//# upgrade [OPTIONS] --package <PACKAGE> --upgrade-capability <UPGRADE_CAPABILITY> --sender <SENDER>
```

#### Options

```
--package <PACKAGE>: the name of the package to upgrade.
--upgrade-capability <UPGRADE_CAPABILITY>: the upgrade capability object that authorizes the upgrade.
--dependencies <DEPENDENCIES> (optional): a list of dependencies required for the upgraded package.
--sender <SENDER>: the account that submits the transaction.
--gas-budget <GAS_BUDGET>: the maximum amount of gas allowed for the upgrade transaction.
--syntax <SYNTAX> (optional): specifies the syntax type (source or ir). Defaults to source.
--policy <POLICY> (optional, default: compatible): the upgrade policy:
    compatible – Allows only compatible upgrades.
    additive – Allows adding new functionality but not modifying existing.
    dep_only – Allows only dependency updates.
--gas-price <GAS_PRICE> (optional): specifies the gas price for the transaction.
```

> Example:

Reviewer: ```suggestion
#### Example
```

module test2::test_coin {
    public struct TestCoin has key, store {
        id: UID,
        amount: u64,
    }

    public fun mint() { }
}
```

`.exp` output:

```
processed 3 tasks

init:
acc1: object(0,0)

task 1 'publish'. lines 3-9:
created: object(1,0), object(1,1)
mutated: object(0,0)
gas summary: computation_cost: 1000000, storage_cost: 5958400,  storage_rebate: 0, non_refundable_storage_fee: 0

task 2 'upgrade'. lines 11-19:
created: object(2,0)
mutated: object(0,0), object(1,0)
gas summary: computation_cost: 1000000, storage_cost: 6171200,  storage_rebate: 2622000, non_refundable_storage_fee: 0
```

### `stage-package`

The `StagePackage` subcommand is used to prepare a package for future upgrades by staging it. Staging allows validation of the package's bytecode, dependencies, and structure before committing the upgrade. The package remains unpublished until explicitly upgraded.

#### Syntax

```
//# stage-package [OPTIONS]
```

#### Options

```
--syntax <SYNTAX>: specifies the syntax type (source or ir).
--dependencies <DEPENDENCIES> (optional): a list of package dependencies required for the staged package.
```

> Example:

Reviewer: ```suggestion
#### Example
```

```move
//# init --addresses test=0x0 test2=0x0 --accounts acc1

//# stage-package
module test::test_coin {
    public struct TestCoin has key, store {
        id: UID,
        amount: u64
    }
}
```

`.exp` output:

```
processed 2 tasks

init:
acc1: object(0,0)
```

### `set-address`

The `SetAddress` subcommand assigns a named address to an existing object, enabling it to be referenced by a human-readable identifier in subsequent commands. This is useful for improving readability and maintainability when working with objects in Move transactions.

#### Syntax

```
//# set-address <NAME> <INPUT>
```

#### Options

```
<NAME>: the human-readable identifier for the address.
<INPUT>: the value to assign to the named address. This can be:
  A Move object (e.g., object(0x123))
  A digest of a staged package (e.g., digest(MyPackage))
  A receiving object (e.g., receiving(0x456))
  A shared immutable object (e.g., immshared(0x789))
```

> Example:

Reviewer: ```suggestion
#### Example
```

```

Forces the creation of 5 checkpoints:

```
//# init --accounts acc1 --simulator

//# create-checkpoint 5

//# view-checkpoint
```

`.exp` output:

```
processed 3 tasks

init:
acc1: object(0,0)

task 1 'create-checkpoint'. lines 3-3:
Checkpoint created: 5

task 2 'view-checkpoint'. lines 5-5:
CheckpointSummary { epoch: 0, seq: 5, content_digest: D3oWLCcqoa1D15gxzvMaDemNNY8YYVspAkYkcmtQKWRt,
            epoch_rolling_gas_cost_summary: GasCostSummary { computation_cost: 0, storage_cost: 0, storage_rebate: 0, non_refundable_storage_fee: 0 }}
```

### `advance-epoch`

The `AdvanceEpoch` subcommand manually advances the epoch in the system. Epochs represent discrete time periods in a network, and transitioning to a new epoch can involve validator set changes, protocol upgrades, and other governance actions.

#### Syntax

```
//# advance-epoch [OPTIONS] [COUNT]
```

#### Options

```
[COUNT]: specifies the number of epochs to advance. If omitted, the command advances by one epoch.
--create-random-state: if set, generates a new random state when advancing the epoch.
```

> Examples:

Reviewer: ```suggestion
#### Examples
```

```

Example:

```move
//# init --protocol-version 1 --simulator

//# create-checkpoint

// advance the clock by 1ms, next checkpoint timestmap should be 1970-01-01T00:00:00:001Z
//# advance-clock --duration-ns 1000000

//# create-checkpoint
```

`.exp` output:

```
processed 23 tasks

task 1 'create-checkpoint'. lines 3-5:
Checkpoint created: 1

task 3 'create-checkpoint'. lines 8-10:
Checkpoint created: 2
```

### `set-random-state`

The `SetRandomState` subcommand sets the blockchain's random state for testing and development purposes. It allows specifying a randomness round, input bytes for randomness, and an initial version number for tracking.

#### Syntax

```
//# set-random-state --randomness-round <RANDOMNESS_ROUND> --random-bytes <RANDOM_BYTES> --randomness-initial-version <RANDOMNESS_INITIAL_VERSION>
```

#### Options

```
--randomness-round <RANDOMNESS_ROUND>: specifies the round number for which the randomness is being set.
--random-bytes <RANDOMNESS_BYTES>: the base64-encoded string representing the new randomness state.
--randomness-initial-version <RANDOMNESS_INITIAL_VERSION>: the version number at which this randomness state is initially set.
```

> Example:

Reviewer: ```suggestion
#### Example
```

                    randomnessRound
                    randomBytes
                    randomnessObjInitialSharedVersion
                }
            }
        }
    }
}
```

`.exp` output:

```
processed 5 tasks

task 1 'create-checkpoint'. lines 3-3:
Checkpoint created: 1

task 3 'create-checkpoint'. lines 7-7:
Checkpoint created: 2

task 4 'run-graphql'. lines 9-24:
Response: {
  "data": {
    "transactionBlocks": {
      "nodes": [
        {
          "kind": {
            "__typename": "RandomnessStateUpdateTransaction",
            "epoch": {
              "epochId": 0
            },
            "randomnessRound": 0,
            "randomBytes": "SGVsbG8gU3Vp",
            "randomnessObjInitialSharedVersion": 2
          }
        }
      ]
    }
  }
}
```

### `view-checkpoint`

The `ViewCheckpoint` subcommand retrieves and displays the latest checkpoint information from the blockchain. This is useful for debugging, monitoring, and ensuring data consistency across nodes.

#### Syntax

```
//# view-checkpoint
```

- Fetches the most recent checkpoint from the blockchain.
- Outputs details such as the checkpoint sequence number, epoch, digest, and gas info.

> Example:

Reviewer: ```suggestion
#### Example
```

```move
//# init --accounts acc1 --simulator

//# view-checkpoint
```

`.exp` output:

```
processed 3 tasks

init:
acc1: object(0,0)

task 1 'view-checkpoint'. lines 3-3:
CheckpointSummary { epoch: 0, seq: 0, content_digest: 3XhwVx9s5eS29WHJN1AUcM4STEZREm2dpSP6nstENKJ2,
            epoch_rolling_gas_cost_summary: GasCostSummary { computation_cost: 0, storage_cost: 0, storage_rebate: 0, non_refundable_storage_fee: 0 }}
```

### `run-graphql`

Allows to execute GraphQL queries with optional options and returns the output.

> #### Syntax:

Reviewer: ```suggestion
#### Syntax
```

```
//# run-graphql [OPTIONS]
<GraphQL-query>
```

> #### Options:

Reviewer: ```suggestion
#### Options
```

        }
        allBalances: balances {
          nodes {
            coinType {
              repr
            }
            coinObjectCount
            totalBalance
          }
        }
      }
    }
  }
}
```

An example of a query that generates an object cursor at runtime:

```
//# run-graphql --cursors @{obj_6_0}
{
  address(address: "@{A}") {
    objects(first: 2 after: "@{cursor_0}") {
      edges {
        node {
          contents {
            json
          }
        }
      }
    }
  }
}
```

### `force-object-snapshot-catchup`

The `ForceObjectSnapshotCatchup` subcommand forces the system to catch up on object snapshots between a specified range of checkpoints. This is useful for ensuring that object state updates are fully synchronized across nodes, particularly in scenarios where snapshots may be lagging behind.

#### Syntax

```
//# force-object-snapshot-catchup -start-cp <START_CP> --end-cp <END_CP>
```

#### Options

```
--start-cp <START_CP>: the starting checkpoint sequence number from which to begin catching up object snapshots.
--end-cp <END_CP>: the ending checkpoint sequence number up to which object snapshots should be caught up.
```

> Example:

Reviewer: ```suggestion
#### Example
```

#### Syntax

```
//# bench [OPTIONS] [NAME]
```

#### Options

```
[NAME]: the name of the function to benchmark. If omitted, the transaction must be explicitly defined through other options. Expects 3 distinct parts - address, module, and struct.
--sender <SENDER>: the account that initiates the transaction. If omitted, the default sender account will be used.
--gas-price <GAS_PRICE>: specifies the gas price for the transaction execution.
--summarize: if set, produces a summarized output of the benchmark results instead of detailed logs.
--signers <SIGNERS>: a list of signers for the transaction, used when executing a function that requires multiple signers.
--args <ARGS>: a list of input arguments passed to the function being benchmarked. Arguments must match the expected input format.
--type-args <TYPE_ARGS>: specifies the type parameters used in the function execution.
--gas-budget <GAS_BUDGET>: sets the maximum amount of gas units allocated for the transaction execution.

> --syntax <SYNTAX>: dfines the Move syntax type for transaction execution, either source (default) or IR.

Reviewer: ```suggestion
--syntax <SYNTAX>: defines the Move syntax type for transaction execution, either source (default) or IR.
```

```

> Example:

Reviewer: ```suggestion
#### Example
```

```move
//# init --addresses test=0x0 --accounts acc1 --protocol-version 1

//# publish --upgradeable --sender acc1
module test::test_coin {
    public struct TestCoin has key, store {
        id: UID,
        amount: u64
    }

    public entry fun test_coin_mint(amount: u64, ctx: &mut TxContext) {
        let id = object::new(ctx);
        let test_coin = TestCoin { id, amount };
        transfer::public_transfer(test_coin, tx_context::sender(ctx));
    }
}

//# bench test::test_coin::test_coin_mint --sender acc1 --args 10000000
```

This benchmarks test_coin_mint, executed by acc1.
Passes amount of coints to mint: 10000000.

`.exp` output:

```
processed 3 tasks

init:
acc1: object(0,0)

task 1 'publish'. lines 3-15:
created: object(1,0), object(1,1)
mutated: object(0,0)
gas summary: computation_cost: 1000000, storage_cost: 7220000,  storage_rebate: 0, non_refundable_storage_fee: 0
```

> ## How `run_test` compares a Move File with the Corresponding .exp file

Reviewer: ```suggestion
## How `run_test` Compares a Move File With the Corresponding `.exp` File
```

The `test_runner` compares `.move` files by executing them and comparing the output with an expected `.exp` files. This ensures that the Move program behaves as expected.

The main entry function for this process is `run_test_impl`.

```rust
pub async fn run_test_impl<'a, Adapter>(
    path: &Path,
    fully_compiled_program_opt: Option<Arc<FullyCompiledProgram>>,
) -> Result<(), Box<dyn std::error::Error>>
where
    Adapter: MoveTestAdapter<'a>,
    Adapter::ExtraInitArgs: Debug,
    Adapter::ExtraPublishArgs: Debug,
    Adapter::ExtraValueArgs: Debug,
    Adapter::ExtraRunArgs: Debug,
    Adapter::Subcommand: Debug,
{
    // Executes the .move file and captures output
    let output = handle_actual_output::<Adapter>(path, fully_compiled_program_opt).await?;
    
    // Compares actual output with expected .exp file
    handle_expected_output(path, output.0)?;

    Ok(())
}
```

> ### **Execution Process in `handle_actual_output`**

Reviewer: ```suggestion
### Execution Process in `handle_actual_output`
```

   - Ensures that the file contains at least one valid command.

3. Executing Each Task and Capturing the Output.
   - `handle_known_task` is responsible for executing parsed tasks from `.move` files based on its type (e.g., `init`, `programmable`, `publish`).

   ```rust
   for task in tasks {
       handle_known_task(&mut output, &mut adapter, task).await;
   }
   ```

   It uses `handle_command` to execute each command:

   - Init: initializes the test environment.
   - Run: calls a Move function.
   - PrintBytecode: compiled Move binary and prints its bytecode instructions.
   - Subcommand: handles other subcommands like `transfer-object`, `create-checkpoint`, etc.

   ```rust
   async fn handle_command(...) {
     match command {
               TaskCommand::Init { .. } => {
                   panic!("The 'init' command is optional. But if used, it must be the first command")
               }
               TaskCommand::Run(run_cmd, args) => { }
               TaskCommand::Publish(run_cmd, args) => { }
               TaskCommand::PrintBytecode(run_cmd, args) => { }
               TaskCommand::Subcommand(run_cmd, args) => { }
     }
   }
   ```

> ### **Verification Process in `handle_expected_output`**

Reviewer: ```suggestion
### Verification Process in `handle_expected_output`
```

> 1. Reading the Expected Output from the Corresponding `.exp` File

Reviewer: ```suggestion
1. Reading the expected output from the Corresponding `.exp` File
```

   - The `.exp` file contains expected execution results for comparison.

2. Comparing Actual and Expected Output

   - The function checks if the produced output matches the expected results.

>    - If the output does not match, it - computes the difference between expected and actual outputs and provides a mechanism for updating baselines if necessary.

Reviewer: ```suggestion
   - If the output does not match, it computes the difference between expected and actual outputs and provides a mechanism for updating baselines if necessary.
```

   ```rust
   if output != expected_output {
           let msg = format!(
               "Expected errors differ from actual errors:\n{}",
               format_diff(expected_output, output),
           );
           anyhow::bail!(add_update_baseline_fix(msg))
       } else {
           Ok(())
       }
   ```

> ### **Structure of the `.move` File.**

Reviewer: ```suggestion
### Structure of the `.move` File
```

A `.move` test file consists of commands and Move code, which are executed step by step. The structure follows these rules:

- Commands start with //#.
- Commands should be separated by an empty line, except when Move code is immediately following a specific command.
- The first command must be init.

Example of `.move` file structure:

```move
//# init --protocol-version 1 --addresses P0=0x0 --accounts A --simulator

// Split off a gas coin, so we have an object to query
//# programmable --sender A --inputs 1000 @A
//> SplitCoins(Gas, [Input(0)]);
//> TransferObjects([Result(0)], Input(1))

//# create-checkpoint

//# run-graphql
{
  sender: owner(address: "@{A}") {
    asObject { digest }
  }

  coin: owner(address: "@{obj_1_0}") {
    asObject { digest }
  }
}
```

> ### **Structure of a `.exp` File**

Reviewer: ```suggestion
### Structure of a `.exp` File
```

> ### **Structure of a `.exp` File**

Reviewer: ```suggestion
### Structure of a `.exp` File
```

> A `.exp` file contains the expected output for the .move test. It includes:

Reviewer: ```suggestion
A `.exp` file contains the expected output for the `.move` test. It includes:
```

- A summary of processed tasks
- Execution results for each task
- Gas usage and storage fees (where applicable)
- GraphQL query responses (if applicable)
- The first line states the number of processed tasks.
- Each task output starts with task index, name, and line range.

Example of `.exp` file structure:

```exp
processed 4 tasks

init:
A: object(0,0)

task 1 'programmable'. lines 8-10:
created: object(1,0)
mutated: object(0,0)
gas summary: computation_cost: 1000000, storage_cost: 1976000,  storage_rebate: 0, non_refundable_storage_fee: 0

task 2 'create-checkpoint'. lines 12-12:
Checkpoint created: 1

task 3 'run-graphql'. lines 14-23:
Response: {
  "data": {
    "sender": {
      "asObject": null
    },
    "coin": {
      "asObject": {
        "digest": "4KjRv4dmBLHXtbw9LJJXeYfSoWeY8aFdkG7FooAqTZWq"
      }
    }
  }
}
```

It includes all 4 tasks execution with their output:

1. Init
2. Programmable
3. Create checkpoint
4. Run graphql

> ### **Extending handle_subcommand and Creating New Subcommands**

Reviewer: ```suggestion
### Extending `handle_subcommand` and Creating New Subcommands
```

The `handle_subcommand` function is responsible for executing subcommands within the test framework. Each subcommand represents a specific action, such as executing Move calls, transferring objects, or publishing Move packages. If you need to extend `handle_subcommand` by adding a new subcommand, follow these steps:

> 1. **Define the New Subcommand in the Enum**

Reviewer: ```suggestion
#### 1. Define the New Subcommand in the Enum
```

New subcommands should be added to the `IotaSubcommand` enum, located inside the test adapter implementation:

```rust
#[derive(Debug)]
pub enum IotaSubcommand<ExtraValueArgs, ExtraRunArgs> {
    // Existing subcommands
    ViewObject(ViewObjectCommand),
    TransferObject(TransferObjectCommand),
    ProgrammableTransaction(ProgrammableTransactionCommand),
    ConsensusCommitPrologue(ConsensusCommitPrologueCommand),
    AdvanceEpoch(AdvanceEpochCommand),
    AdvanceClock(AdvanceClockCommand),
    CreateCheckpoint(CreateCheckpointCommand),
    ForceObjectSnapshotCatchup(ForceObjectSnapshotCatchup),
    SetAddress(SetAddressCommand),
    SetRandomState(SetRandomStateCommand),
    RunGraphql(RunGraphqlCommand),

    // New Subcommand
    CustomObjectAction(CustomObjectActionCommand),
}
```

> 2. **Define the Command Struct**

Reviewer: ```suggestion
#### 2. Define the Command Struct
```

Each subcommand requires a struct that defines its arguments and expected input parameters. The struct should include:

> Named fields for each argument.
> #[derive(Debug)] for logging and debugging.

Reviewer: ```suggestion
* Named fields for each argument.
* #[derive(Debug)] for logging and debugging.
```

```rust
#[derive(Debug)]
pub struct CustomObjectActionCommand {
    pub target: String,  // Example argument
    pub value: u64,
    pub args: Vec<SomeArgs>
}
```

This struct will be parsed and used when executing the subcommand.

> 3. **Implement the Logic for the Subcommand**

Reviewer: ```suggestion
#### 3. Implement the Logic for the Subcommand
```

Modify the `handle_subcommand` function inside `IotaTestAdapter` to include the new subcommand's logic.

Locate the match statement inside `handle_subcommand`, and add your new subcommand:

```rust
async fn handle_subcommand(
    &mut self,
    task: TaskInput<Self::Subcommand>,
) -> anyhow::Result<Option<String>> {
    self.next_task();
    
    match command {
        // Other commands handling
        // ...

        // Custom subcommand implementation
        IotaSubcommand::CustomObjectAction(cmd) => {
            // Logic
        }
    }
}
```

> 4. **Add a Test Case** - `.move` and `.exp` files to test different scenarios.

Reviewer: ```suggestion
#### 4. Add Tests Cases
Create `.move` and `.exp` files to test different scenarios.
```
